### PR TITLE
CI: Add miltilibbed Elf32 & glibc configs for ARC

### PR DESCRIPTION
**Pull request by abrodkin**
As of today baremetal (AKA "Elf32") & Linux glibc toolchains are even more important than Linux uClibc one for ARC, so adding them.

@cpackham if +2 more full multilib toolchains for just one architecture is way too much given amount of resources we have for per-PR testing here, I may drop existing uClibc config.

### CODE REVIEW

**Comment by abrodkin** [comment]
Oops, forgot to add off-the-tree patch for glibc on ARCompact (ARC700), will do now!

**Comment by abrodkin** [comment]
Ok, we now need https://github.com/crosstool-ng/crosstool-ng/pull/1457 merged first to get a clean build of `arc-multilib-linux-gnu`.

**Comment by cpackham** [comment]
> As of today baremetal (AKA "Elf32") & Linux glibc toolchains are even more important than Linux uClibc one for ARC, so adding them.
> 
> @cpackham if +2 more full multilib toolchains for just one architecture is way too much given amount of resources we have for per-PR testing here, I may drop existing uClibc config.

I think we're OK at the moment. If github start to complain about us using too much CI resource we might have to make some choices about which configs get CI but seems OK for now.

**Comment by cpackham** [comment]
@abrodkin could you rebase this on top of master now that I've merged the ARC700 PR.

**Comment by abrodkin** [comment]
Hm, that I don't understand while building on Mac:
```
(macos-10.15, arc-multilib-linux-gnu)
failed 32 minutes ago in 56m 48s 

[INFO ]  Installing C library
[INFO ]    =================================================================
[INFO ]    Building for multilib 1/7: ''
[EXTRA]      Cleaning up start files
[EXTRA]      Configuring C library
[EXTRA]      Building C library
Error:       ../sysdeps/arc/fpu/fegetround.c:28:19: error: '__FPU_RND_SHIFT' undeclared (first use in this function)
Error:       ../sysdeps/arc/fpu/fegetround.c:28:38: error: '__FPU_RND_MASK' undeclared (first use in this function)
Error:       ../sysdeps/arc/fpu/fegetround.c:29:1: error: control reaches end of non-void function [-Werror=return-type]
Error:       make[3]: *** [/Users/runner/work/crosstool-ng/crosstool-ng/.build/arc-multilib-linux-gnu/build/build-libc-final/multilib/sysd-rules:417: /Users/runner/work/crosstool-ng/crosstool-ng/.build/arc-multilib-linux-gnu/build/build-libc-final/multilib/math/fegetround.o] Error 1
Error:       make[3]: *** Waiting for unfinished jobs....
Error:       ../sysdeps/arc/fpu/fesetround.c:29:17: error: '__FPU_RND_SHIFT' undeclared (first use in this function)
Error:       ../sysdeps/arc/fpu/fesetround.c:29:36: error: '__FPU_RND_MASK' undeclared (first use in this function)
Error:       make[3]: *** [/Users/runner/work/crosstool-ng/crosstool-ng/.build/arc-multilib-linux-gnu/build/build-libc-final/multilib/sysd-rules:417: /Users/runner/work/crosstool-ng/crosstool-ng/.build/arc-multilib-linux-gnu/build/build-libc-final/multilib/math/fesetround.o] Error 1
Error:       make[2]: *** [Makefile:479: math/others] Error 2
Error:       make[1]: *** [Makefile:9: all] Error 2
```

I'm pretty sure all worked well on Linux locally. Will give it a try now again.

**Comment by abrodkin** [comment]
Hm, and indeed even here in the CI that problem doesn't happen, see https://github.com/crosstool-ng/crosstool-ng/runs/1902650827: toolchains (ubuntu-latest, arc-multilib-linux-gnu) succeeded 7 hours ago in 1h 23m 7s.

But we'll get to the bottom of that :)

**Comment by cpackham** [comment]
We have hit some odd failures on macOS for mips64 as well. I don't have access to a macOS system to test with so I can't help much. @galak any pointers?

If all else fails we can add the arc configurations to the excludes for macOS.

**Comment by abrodkin** [comment]
@cpackham while that problem observed only on Mac is not expected (see Ubuntu builds are all good and reported failure shouldn't happen at all as on ARC700 there's no FPU and so we shouldn't even reach reported location in code) I'm wondering if there's actually any value in building Linux toolchains (i.e uClibc & glibc) for/on Mac?

Bare-metal toolchains for both Windows & macOS are pretty much self-contained, so there's a real use for them.
Linux toolchains are more of a starting point for building something much bigger with say Buildroot or OpenEmbedded/Yocto as there're way too many libs and apps you may need to have in your custom self-built Linux distro. And then I guess neither Buildroot nor OE are usable on Windows (we're not talking about WSL here) or macOS.

In that sense maybe not even try to do that exercise of building Linux toolchains on/for Mac?

FWIW I don't have access to really powerful Mac machines well-suitable for building toolchains, so I use canadian-cross for building for Mac. And while Elf toolchains build perfectly fine, glibc one I just tried failed much earlier on building `gettext` for the host:
```
[ALL  ]    In file included from /osxcross/SDK/MacOSX10.10.sdk/usr/include/dispatch/dispatch.h:51,
[ALL  ]                     from /osxcross/SDK/MacOSX10.10.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:15,
[ALL  ]                     from /osxcross/SDK/MacOSX10.10.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:13,
[ALL  ]                     from .../.build/HOST-x86_64-apple-darwin14/arc-snps-linux-gnu/src/gettext/gettext-runtime/intl/langprefs.c:33:
[ERROR]    /osxcross/lib/gcc/x86_64-apple-darwin14/10.2.0/include-fixed/dispatch/object.h:373:3: error: unknown type name 'dispatch_block_t'; did you mean 'dispatch_object_t'?
[ALL  ]      373 |   dispatch_block_t notification_block);
[ALL  ]          |   ^~~~~~~~~~~~~~~~
[ALL  ]          |   dispatch_object_t
[ALL  ]    In file included from /osxcross/SDK/MacOSX10.10.sdk/usr/include/dispatch/dispatch.h:51,
[ALL  ]                     from /osxcross/SDK/MacOSX10.10.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:15,
[ALL  ]                     from /osxcross/SDK/MacOSX10.10.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:13,
[ALL  ]                     from .../.build/HOST-x86_64-apple-darwin14/arc-snps-linux-gnu/src/gettext/gettext-runtime/intl/setlocale.c:52:
[ERROR]    /osxcross/lib/gcc/x86_64-apple-darwin14/10.2.0/include-fixed/dispatch/object.h:373:3: error: unknown type name 'dispatch_block_t'; did you mean 'dispatch_object_t'?
[ALL  ]      373 |   dispatch_block_t notification_block);
[ALL  ]          |   ^~~~~~~~~~~~~~~~
[ALL  ]          |   dispatch_object_t
```

I'll try to rebuild `osxcross` with more up-to-date MacOS SDK in a hope it works any better :)

**Comment by cpackham** [comment]
Yeah we did [discuss ]([https://github.com/crosstool-ng/crosstool-ng/pull/1418#issuecomment-726037819) that the use-case for macOS was probably more around bare-metal than embedded linux. But then there are a few people tinkering with raspberry pi using a mac who just want to build a few applications to add to the existing system. 

As I said I'm happy to exclude the `arc-*-linux-gnu` and `arc-*-linux-uclibc` on macOS. No point in putting too much effort into getting it working.

**Comment by abrodkin** [comment]
Ok then, indeed let's not waste time on something not yet being used. If there's a need for that we'll reenable what's needed.

**Comment by abrodkin** [comment]
@cpackham well maybe to speed-up CI reverse filtering logic and only add Elf toolchains + ARM Linux one for Mac builds?

**Comment by cpackham** [comment]
> @cpackham well maybe to speed-up CI reverse filtering logic and only add Elf toolchains + ARM Linux one for Mac builds?

Not a bad idea. I also want to tackle #808 at some point which will speed things up a lot. Lets get arc out of the way then worry about speeding things up later.

**Review by cpackham — APPROVED** [review]
I'll just wait for the CI run to finish on this and will merge after that.

**Comment by abrodkin** [comment]
@cpackham seems like my change doesn't really affect exclusion of samples as `arc-multilib-linux-gnu` keeps building on MacOS :(
What did I do wrong?

**Comment by cpackham** [comment]
I think it might need to be entries of host/sample pairs not an array.

**Comment by abrodkin** [comment]
@cpackham that seems to be the case, see https://github.community/t/improve-matrix-exclusion-feature/16340.
Let's hope it works finally :)

**Review by cpackham — APPROVED** [review]

**Pull request merged**
